geti_sdk/deployment/inference_hook_interfaces.py

```diff
@@ -0,0 +1,263 @@
+    Base interface class for post inference triggers, actions and hooks
+    """
+
+    _override_from_dict_: bool = False
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Create a new PostInferenceObject instance
+        """
+        instance = super().__new__(cls)
+        instance._argument_dict_ = {"args": args, "kwargs": kwargs}
+        return instance
+
+    def __init__(self):
+        self._constructor_arguments_ = self.__get_constructor_arguments()
+
+    def __get_constructor_arguments(self) -> Dict[str, Any]:
+        """
+        Return the arguments used for constructing the PostInferenceAction object
+
+        :return: Dictionary containing the constructor parameter names as keys, and
+            parameter values as values
+        """
+        constructor_argument_params = inspect.signature(self.__init__).parameters
+        parameters: Dict[str, Any] = {}
+        args = self._argument_dict_.get("args", ())
+        kwargs = self._argument_dict_.get("kwargs", {})
+        for index, (pname, parameter) in enumerate(constructor_argument_params.items()):
+            if index + 1 <= len(args):
+                parameters.update({pname: args[index]})
+            elif pname in kwargs.keys():
+                parameters.update({pname: kwargs[pname]})
+            else:
+                parameters.update({pname: parameter.default})
+        return parameters
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Return a dictionary representation of the PostInferenceObject
+
+        :return: Dictionary representing the class name and its constructor parameters
+        """
+        constructor_args: Dict[str, Any] = {}
+        for key, value in self._constructor_arguments_.items():
+            if isinstance(value, PostInferenceObject):
+                constructor_args.update({key: value.to_dict()})
+            else:
+                constructor_args.update({key: value})
+        return {type(self).__name__: constructor_args}
+
+    @classmethod
+    @abstractmethod
+    def from_dict(cls, input_dict: Dict[str, Any]) -> "PostInferenceObject":
+        """
+        Construct a PostInferenceObject from an input dictionary `input_dict`
+
+        :param input_dict: Dictionary representation of the PostInferenceObject
+        :return: Instantiated PostInferenceObject, according to the input dictionary
+        """
+        return NotImplemented
+
+
+class PostInferenceTrigger(PostInferenceObject):
+    """
+    Base class for post inference triggers
+
+    Post inference triggers are used in inference hooks. They define a condition
+    (or set of conditions) that determines whether any post-inference action should be
+    taken.
+
+    :param threshold: Threshold to use in activating the trigger. If the computed
+        score is above the threshold, the trigger emits a positive decision and any
+        downstream actions will be performed
+    """
+
+    def __init__(self, threshold: float = 0.5):
```

Comment:
`threshold` parameter here looks excessive. Prediction confidence is just one of many possible parameters and only part of sub-classes will use it.

Reply:
Oh, I see, it is not a confidence threshold, but rather some abstract threshold.

Reply:
Yeah it's a threshold for the trigger score, the exact function depends on the trigger implementation